### ISSUE TITLE
Create Solr facets for search filters

### ISSUE THREAD

**Issue by paynejd** [OpenConceptLab/oclapi#71]
Each resource has metadata that can be used to filter search results or to provide options to users when editing/creating entries. For example, concept metadata includes concept class, datatype, name locale, name type, description locale, and description type. 

Currently we have no way of dynamically populating a list of existing options for any of these metadata fields with the option to filter these options based on some other value. For example, you can query the list of all concept name types used in OCL to dynamically populate a dropdown when a user is creating a new concept name. Or, you can query the list of concept classes used within a specific source when dynamically creating the list of concept class filters on the source page.

Implement facets for the following fields:
* **Concepts:** class, datatype, concept name locale, concept name type, concept description locale, concept description type
* **Sources:** type, locale
* **Collections:** type, locale
* **Mappings:** type

Originally estimated at 4 hours of work.

**Comment by MisterNando** [comment]
Estimates:

Original - 4h
Updated - 4h

**Comment by MisterNando** [comment]
@paynejd , @pkshiu , here's how I'm envisioning how results & facets will be returned.  Note that the JSON coming back would now be an object as opposed to a list:

```
{
facets: {
fields: {
language: [3]
0:  [2]
0:  "en"
1:  4
-
1:  [2]
0:  "es"
1:  1
-
2:  [2]
0:  "fr"
1:  1
-
-
}-
dates: {}
queries: {}
}-
results: [8]
0:  {
short_code: "LOINC"
name: " Logical Observation Identifiers Names and Codes"
url: "/orgs/REG/sources/LOINC/"
owner: "REG"
owner_type: "Organization"
owner_url: "/orgs/REG/"
}-
1:  {
short_code: "SRC"
name: "Johnny's Test Source"
url: "/users/johnnytest/sources/SRC/"
owner: "johnnytest"
owner_type: "User"
owner_url: "/users/johnnytest/"
}-
2:  {
short_code: "SNOMED"
name: "SNOMED"
url: "/orgs/WHO/sources/SNOMED/"
owner: "WHO"
owner_type: "Organization"
owner_url: "/orgs/WHO/"
}-
3:  {
short_code: "open"
name: "Open"
url: "/users/johnnytest/sources/open/"
owner: "johnnytest"
owner_type: "User"
owner_url: "/users/johnnytest/"
}-
4:  {
short_code: "NEW"
name: "New Source"
url: "/users/johnnytest/sources/NEW/"
owner: "johnnytest"
owner_type: "User"
owner_url: "/users/johnnytest/"
}-
5:  {
short_code: "ImportTest"
name: "Source for testing import"
url: "/users/johnnytest/sources/ImportTest/"
owner: "johnnytest"
owner_type: "User"
owner_url: "/users/johnnytest/"
}-
6:  {
short_code: "CIELTEST"
name: "CIELTEST"
url: "/orgs/MCL/sources/CIELTEST/"
owner: "MCL"
owner_type: "Organization"
owner_url: "/orgs/MCL/"
}-
7:  {
short_code: "CIELTEST"
name: "CIELTEST"
url: "/users/johnnytest/sources/CIELTEST/"
owner: "johnnytest"
owner_type: "User"
owner_url: "/users/johnnytest/"
}-
-
}
```

**Comment by MisterNando** [comment]
Regarding the above, you can ignore facets.dates, facets.queries (those come with the response from Solr) and the escape characters (those come from the tool I used to pretty-print the JSON).

**Comment by MisterNando** [comment]
@paynejd , @pkshiu - awaiting feedback on proposed new result format (object vs. list - see above) so I can move forward with this.

**Comment by paynejd** [comment]
@MisterNando - Are you recommending that we change all API responses or only lists? Is it possible to only change the format if there is a flag in the parameter that requests facets/object? This could work well because we only need to take advantage of facets on the screens in the search screens in the web app. I hate to complicate the spec, but I also hate to change the entire spec. What are your thoughts on this approach?

@pkshiu - Can you estimate effort to switch the search apps to the object result vs. switching all API results over to the object result?

**Comment by MisterNando** [comment]
I'm proposing this format for all list APIs.

**Comment by paynejd** [comment]
@MisterNando - Changing this for all list formats is a significant departure from the spec. Let's keep the current list format and then return the object/facet info only if a header attribute is set. e.g. return_object=true

**Comment by MisterNando** [comment]
Specifying a flag as a header attribute seems like a departure from the
spec as well.  Can you think of a less intrusive way to express the facets?

**Comment by pkshiu** [comment]
@MisterNando @paynejd have we come to a conclusion on this? I am not really following the object idea. We really just want a list of concept_types, source locales, etc. Each with it's own endpoint for the retrieval, right?

**Comment by paynejd** [comment]
Changing the output for all lists is a significant change to the *existing*
specs, which would also require significant rewrites to the web app. In
addition, facets are only useful when using an interactive search
interface, so it makes sense for the list format to be the default for the
API and the object format optional. Formats are typically specified in the
request header as opposed to the URL parameters, which is why I suggested
this.

Unless there is a good reason not to do it this way, it makes the most
sense to me.

**Comment by MisterNando** [comment]
I just think you're overgeneralizing.  The problem we are trying to solve
is "where do facets fit into the search/filtering functionality?" - not
"how can we accommodate a different response format?"  Can we solve the
former problem without having to address to latter?

**Comment by paynejd** [comment]
We know where they fit into the functionality, we are just not clear on the
format for returning them. Changing thee response format of all lists is
unacceptable unless there is an extremely good reason to do so. If you do
not think my proposal is a good option then please propose an alternative
that does not break what has already been implemented.

**Comment by MisterNando** [comment]
Jon, I'm trying to gather information so I can answer that question.  What
are the use cases for facets?  Are they the exception, or are they the
rule?  If they're an exception, then fine - let's put a flag in the request
header.  But let's call it what it is - i.e. "returnFacets" - rather than
overgeneralize.  Maybe the facets should come back in the header too -
would they fit?  Or maybe there should be 2 separate calls - one for the
result list, and one for the facets - that almost feels cleaner (since
faceting is more of an application feature than an API feature).

**Comment by paynejd** [comment]
The "known" use cases for facets are the interactive search tools, which
means they will be used by the web app for "global search", and for search
within specific orgs/users/sources/collections. One other defined use case
(that is not currently contracted out to anyone) is a dynamic widget that
others could integrate into their website to search resources. There may be
other use cases, but these are the only ones on the radar for now. In
situations in which a 3rd party is using the API directly, unless they have
re-created an interactive search tool, they will most likely need to know
enough about the data to specify the correct filters/facets in advance and
will not be able to take advantage of facet information included in the
response -- meaning, they will probably use the list format.

Regarding the options, seems like there 4 right now:

1) Return facets in the response header -- if we can be sure that they will
always fit
2) Split the facets into a separate request (using the header parameter to
the same endpoint with the same filters/facets applied?)
3) Default to list format, and use header parameter to request a response
format that includes the facets (the object)
4) Switch all lists over to the object format that always include facets

Regarding including in the response header (#1), there is a lot of
variability between http servers and web browsers to take into
consideration. Take a look at these for a start:

   -
   http://stackoverflow.com/questions/3326210/can-http-headers-be-too-big-for-browsers
   -
   http://stackoverflow.com/questions/1097651/is-there-a-practical-http-header-length-limit
   - http://stackoverflow.com/questions/686217/maximum-on-http-header-values

From my perspective, #4 seems inconsistent with the rest of the API ---
i.e. why not switch *everything* over to objects? -- and also would take a
lot of extra effort on the web application side. #1 seems like we'd always
have to be worrying about peculiar limitations of specific server platforms
and browsers -- what happens when the facet info is over a size limit? #2
is a good option, with the one significant drawback that the query
basically has to be executed twice. #3 is still my preference, but let me
know your thoughts.

Also, let me know if there are any other questions that I can answer.



--
Jonathan Payne, MS
615.579.5413
skype: jonathandavidpayne

**Comment by paynejd** [comment]
Aaron, let me know your thoughts on the above and if you have any additional questions.

**Comment by MisterNando** [comment]
Sure, I'm okay with Option 3.  What would you like the header parameter to be?

**Comment by paynejd** [comment]
Can do "includeFacets: true"?
